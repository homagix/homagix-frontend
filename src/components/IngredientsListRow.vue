<script setup lang="ts">
import type { Ingredient } from "@/types"
import EditableField from "./EditableField.vue"
import AutocompleteField from "./AutocompleteField.vue"
import store from "@/store"

const props = defineProps<{
  ingredient: Ingredient
  editable: boolean
  deletable?: boolean
}>()

const emit = defineEmits<{
  (event: "input", value: Ingredient): void
  (event: "delete", value: Ingredient): void
}>()

const setAmount = (amount: string) => emit("input", { ...props.ingredient, amount: +amount })
const setUnit = (event: Event) => emit("input", { ...props.ingredient, unit: (event.target as HTMLInputElement).value })
const setName = ({ id, name }: { id: string; name: string }) => emit("input", { ...props.ingredient, id, name })
const deleteItem = () => emit("delete", props.ingredient)
</script>

<template>
  <div :class="{ row: true, editable }">
    <EditableField type="number" :value="ingredient.amount" @input="setAmount" :editable="editable" />

    <select :value="ingredient.unit" @input="setUnit" :disabled="!!ingredient.id">
      <option v-for="entry in store.state.units" :key="entry">{{ entry }}</option>
    </select>

    <EditableField v-if="ingredient.id" :value="ingredient.name" :editable="false" />
    <AutocompleteField
      v-else
      id="item-name"
      :value="ingredient.name"
      :list="store.state.ingredients"
      @change="setName"
    />

    <span v-if="deletable" class="icon delete" @click="deleteItem"></span>
  </div>
</template>

<style scoped lang="scss">
.row {
  span {
    white-space: nowrap;
  }

  select {
    border-color: transparent;
    font-size: 100%;
    color: black;

    &:disabled {
      appearance: none;
      padding-right: 15px;
    }
  }

  span.icon.delete {
    display: inline-block;
    margin: 4px 3px;
  }
}
</style>
