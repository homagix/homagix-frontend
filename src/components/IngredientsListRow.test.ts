import { DOMWrapper, mount, shallowMount } from "@vue/test-utils"
import { describe, expect, it } from "vitest"
import IngredientsListRow from "@/components/IngredientsListRow.vue"

const newIngredient = { id: "", amount: 3, name: "def", unit: "Stk" }
const existingIngredient = { id: "123", amount: 2, name: "abc", unit: "ml" }

describe("IngredientsListRow", () => {
  describe("in view mode", () => {
    it("should render", () => {
      const wrapper = shallowMount(IngredientsListRow, { props: { ingredient: existingIngredient, editable: false } })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe("in edit mode", () => {
    it("should render in editable mode", () => {
      const wrapper = mount(IngredientsListRow, { props: { ingredient: existingIngredient, editable: true } })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render a delete button if requested", () => {
      const wrapper = mount(IngredientsListRow, {
        props: { ingredient: existingIngredient, editable: true, deletable: true },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    describe("with existing items", async () => {
      const wrapper = mount(IngredientsListRow, {
        props: { ingredient: existingIngredient, editable: true, deletable: true },
      })
      const [amountField, nameField] = wrapper.findAll("span") as DOMWrapper<HTMLElement>[]

      it("should emit an 'input' event when the amount is modified", async () => {
        amountField.element.innerText = "5"
        await amountField.trigger("input")
        expect(wrapper.emitted().input).toEqual([[{ ...existingIngredient, amount: 5 }]])
      })

      it("should not allow to edit the unit field", () => {
        expect(wrapper.find("select").element.disabled).toBe(true)
      })

      it("should not let the name be edited", () => {
        expect(nameField.element.attributes.getNamedItem("contentEditable")?.value).toBe("false")
      })

      it("should allow to remove items", async () => {
        await wrapper.find(".delete").trigger("click")
        expect(wrapper.emitted().delete).toEqual([[existingIngredient]])
      })
    })

    describe("with a new item", () => {
      const wrapper = mount(IngredientsListRow, { props: { ingredient: newIngredient, editable: true } })
      const unitField = wrapper.find("select")

      it("should show editable fields", () => {
        expect(wrapper.html()).toMatchSnapshot()
      })

      it("should allow to select the unit", () => {
        expect(unitField.element.disabled).toBe(false)
      })

      it("should emit an 'input' event when the unit is modified", async () => {
        unitField.element.selectedIndex = 2
        unitField.trigger("input")
        expect(wrapper.emitted().input).toHaveLength(1)
        expect(wrapper.emitted().input[0]).toEqual([{ id: "", amount: 3, name: "def", unit: "ml" }])
      })
    })
  })
})
