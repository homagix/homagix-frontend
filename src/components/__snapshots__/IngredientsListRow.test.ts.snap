// Vitest Snapshot v1

exports[`IngredientsListRow > in edit mode > should render a delete button if requested 1`] = `
"<div class=\\"row editable\\" data-v-0f750a6a=\\"\\"><span class=\\"number\\" contenteditable=\\"true\\" data-v-09beb9f6=\\"\\" data-v-0f750a6a=\\"\\">2</span><select disabled=\\"\\" data-v-0f750a6a=\\"\\">
    <option data-v-0f750a6a=\\"\\">Stk</option>
    <option data-v-0f750a6a=\\"\\">g</option>
    <option data-v-0f750a6a=\\"\\">ml</option>
    <option data-v-0f750a6a=\\"\\">Bund</option>
    <option data-v-0f750a6a=\\"\\">Glas</option>
  </select><span class=\\"\\" contenteditable=\\"false\\" data-v-09beb9f6=\\"\\" data-v-0f750a6a=\\"\\">abc</span><span class=\\"icon delete\\" data-v-0f750a6a=\\"\\"></span></div>"
`;

exports[`IngredientsListRow > in edit mode > should render in editable mode 1`] = `
"<div class=\\"row editable\\" data-v-0f750a6a=\\"\\"><span class=\\"number\\" contenteditable=\\"true\\" data-v-09beb9f6=\\"\\" data-v-0f750a6a=\\"\\">2</span><select disabled=\\"\\" data-v-0f750a6a=\\"\\">
    <option data-v-0f750a6a=\\"\\">Stk</option>
    <option data-v-0f750a6a=\\"\\">g</option>
    <option data-v-0f750a6a=\\"\\">ml</option>
    <option data-v-0f750a6a=\\"\\">Bund</option>
    <option data-v-0f750a6a=\\"\\">Glas</option>
  </select><span class=\\"\\" contenteditable=\\"false\\" data-v-09beb9f6=\\"\\" data-v-0f750a6a=\\"\\">abc</span>
  <!--v-if-->
</div>"
`;

exports[`IngredientsListRow > in edit mode > with a new item > should show editable fields 1`] = `
"<div class=\\"row editable\\" data-v-0f750a6a=\\"\\"><span class=\\"number\\" contenteditable=\\"true\\" data-v-09beb9f6=\\"\\" data-v-0f750a6a=\\"\\">3</span><select data-v-0f750a6a=\\"\\">
    <option data-v-0f750a6a=\\"\\">Stk</option>
    <option data-v-0f750a6a=\\"\\">g</option>
    <option data-v-0f750a6a=\\"\\">ml</option>
    <option data-v-0f750a6a=\\"\\">Bund</option>
    <option data-v-0f750a6a=\\"\\">Glas</option>
  </select>
  <div class=\\"autocomplete\\" id=\\"item-name\\" value=\\"def\\" data-v-46e78e1c=\\"\\" data-v-0f750a6a=\\"\\"><input type=\\"text\\" role=\\"combobox\\" autocomplete=\\"off\\" autocapitalize=\\"off\\" autocorrect=\\"off\\" spellcheck=\\"false\\" aria-autocomplete=\\"list\\" aria-haspopup=\\"listbox\\" aria-owns=\\"item-name-result-list\\" aria-expanded=\\"false\\" data-v-46e78e1c=\\"\\"><select role=\\"listbox\\" id=\\"item-name-result-list\\" size=\\"2\\" data-v-46e78e1c=\\"\\"></select></div>
  <!--v-if-->
</div>"
`;

exports[`IngredientsListRow > in view mode > should render 1`] = `
"<div class=\\"row\\" data-v-0f750a6a=\\"\\">
  <editable-field-stub type=\\"number\\" editable=\\"false\\" value=\\"2\\" data-v-0f750a6a=\\"\\"></editable-field-stub><select disabled=\\"\\" data-v-0f750a6a=\\"\\">
    <option data-v-0f750a6a=\\"\\">Stk</option>
    <option data-v-0f750a6a=\\"\\">g</option>
    <option data-v-0f750a6a=\\"\\">ml</option>
    <option data-v-0f750a6a=\\"\\">Bund</option>
    <option data-v-0f750a6a=\\"\\">Glas</option>
  </select>
  <editable-field-stub editable=\\"false\\" value=\\"abc\\" data-v-0f750a6a=\\"\\"></editable-field-stub>
  <!--v-if-->
</div>"
`;
